sw/device/silicon_creator/lib/cert/asn1.c

```diff
@@ -0,0 +1,172 @@
+  state->buffer = 0;
+  state->size = 0;
+  state->offset = 0;
+  return kErrorOk;
+}
+
+rom_error_t asn1_push_byte(asn1_state_t *state, uint8_t byte) {
+  return asn1_push_bytes(state, &byte, 1);
+}
+
+rom_error_t asn1_push_bytes(asn1_state_t *state, const uint8_t *bytes,
+                            size_t size) {
+  // TODO: check overflow?
+  if (state->offset + size > state->size)
+    return kErrorAsn1BufferExhausted;
+  memcpy(state->buffer + state->offset, bytes, size);
+  state->offset += size;
+  return kErrorOk;
+}
+
+rom_error_t asn1_start_tag(asn1_state_t *state, asn1_tag_t *new_tag, uint8_t id,
+                           size_t size_hint) {
+  // TODO: use the size hint
+  new_tag->state = state;
+  RETURN_IF_ERROR(asn1_push_byte(state, id));
+  new_tag->len_offset = state->offset;
+  // We do not yet known how many bytes we need to encode the length. For now
+  // reserve three bytes so we can use the 16-bit encoding. This is then fixed
+  // in asn1_finish_tag.
+
+  // Three bytes: one to hold the number of octets to use
+  // and two to store the length.
+  RETURN_IF_ERROR(asn1_push_byte(state, 0x82));
+  RETURN_IF_ERROR(asn1_push_byte(state, 0));
+  RETURN_IF_ERROR(asn1_push_byte(state, 0));
+  new_tag->len_size = 3;
+  return kErrorOk;
+}
+
+rom_error_t asn1_finish_tag(asn1_tag_t *tag) {
+  // Robustness check against accidentely reusing a finished tag.
+  if (tag->state == NULL)
+    return kErrorAsn1Internal;
+  // Compute actually used length.
+  size_t length = tag->state->offset - tag->len_offset - tag->len_size;
+  // Make sure the length is not too large.
+  if (length > 0xffff)
+    return kErrorAsn1Internal;
+  // If the length requires two bytes, then there is no need to move the data.
+  if (length > 0x7f) {
+    // Sanity check: asn1_start_tag should have outputted three bytes.
+    if (tag->len_size != 3)
+      return kErrorAsn1Internal;
+    // Fixup length.
+    tag->state->buffer[tag->len_offset + 0] = 0x82;
+    tag->state->buffer[tag->len_offset + 1] = (uint8_t)(length >> 8);
+    tag->state->buffer[tag->len_offset + 2] = (uint8_t)(length & 0xff);
+  } else {
+    // We only need one byte to hold the length but we allocated three:
+    // we need to shift all the tag data by two bytes. A forward copy
+    // works for that purpose.
+    // TODO: use memmove when we have an implementation.
+    for (size_t i = 0; i < length; i++) {
+      tag->state->buffer[tag->len_offset + 1 + i] =
+          tag->state->buffer[tag->len_offset + tag->len_size + i];
+    }
+    // Fixup length.
+    tag->state->buffer[tag->len_offset] = (uint8_t)length;
+    // Fixed up state offset.
+    tag->state->offset -= tag->len_size - 1;
+  }
+  // Hardening: clear out the tag structure to prevent accident reuse.
+  tag->state = NULL;
+  tag->len_offset = 0;
+  tag->len_size = 0;
+  return kErrorOk;
+}
+
+rom_error_t asn1_push_bool(asn1_state_t *state, bool value) {
+  // Avoid using the tag functions since everything is constant.
```

Comment:
just a question: have you observed poorly optimised code if you used the tag functions?

Reply:
No, I wrote the code early one. I have changed it to use a tag function now for consistency.